_posts/2020-10-24-text-layout.md

```diff
@@ -0,0 +1,217 @@
+
+While Core Text (see below) exposes a hierarchy of objects, DirectWrite uses the [TextLayout](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nn-dwrite-idwritetextlayout) as the primary interface, and exposes internal structure (even including lines) by iterating over a callback per run in the confusingly named [Draw](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritetextlayout-draw) method. The granularity of this callback is a [glyph run](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_glyph_run), which corresponds to "script" in the hierarchy above. Cluster information is provided in an associated [glyph run description](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_glyph_run_description) structure.
+
+There are other ways to access lower level text layout capabilities, including 
+[TextAnalyzer](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nn-dwrite-idwritetextanalyzer), which computes BiDi and line break opportunities, script runs, and shaping. In fact, the various methods on that interface represents much of the internal structure of the text layout engine. Itemization, however, is done in the [FontFallback](https://docs.microsoft.com/en-us/windows/win32/api/dwrite_2/nn-dwrite_2-idwritefontfallback) interface, which was added later.
+
+### Core Text
+
+Another high quality implementation is [Core Text]. I don't personally find it as well designed as DirectWrite, but it does get the job done. In general, though, Core Text is considered a lower level interface, and applications are recommended to use a higher level mechanism (Cocoa text on macOS, Text Kit on iOS).
+
+When doing text layout on macOS, it's probably better to use the platform-provided itemization method (), rather than getting the font list and doing itemization in the client. See [linebender/skribo#14](https://github.com/linebender/skribo/issues/14) for more information on this tradeoff.
+
+### Druid/Piet
+
+At this point, the Druid GUI toolkit does not have its own native text layout engine, but rather *does* provide a [cross-platform API][Piet text layout API] which is delegated to platform text layout engines, DirectWrite and Core Text in particular.
+
+The situation on Linux is currently unsatisfactory, as it's based on the Cairo toy text API. There is work ongoing to improve this, but no promises when.
+
+While the Piet text API is currently fairly basic, I do think it's a good starting point for text layout, especially in the Rust community. While the complexity of Web text basically forces browsers to do all their text layout from scratch, for UI text there are serious advantages to using the platform.
```

Comment:
I’m not sure what you’re saying here in the second sentence.

Reply:
I've expanded it, hopefully it's clear now.